palindrom-error-catcher.html

```diff
@@ -0,0 +1,172 @@
+            </a>
+            <a href="javascript:;" class="reload-btn">Reload now</a>
+        </div>
+    </div>
+    <div class="box-container generic-error-pane" hidden>
+        <div class="box error tappable">
+            <strong>⚠️ An error has occurred.</strong> See console for details.
+            <span>Click here to reload</span>
+        </div>
+    </div>
+</template>
+
+<script>
+    (function () {
+        const template = document.currentScript.ownerDocument.querySelector('#palindrom-error-catcher');
+
+        function bindCallbacks(to) {
+            to.handleConnectionError = to.handleConnectionError.bind(to);
+            to.handleReconnectionCountdownTick = to.handleReconnectionCountdownTick.bind(to);
+            to.handleReconnectionEnd = to.handleReconnectionEnd.bind(to);
+            to.handleGenericError = to.handleGenericError.bind(to);
+            to.cancelReloading = to.cancelReloading.bind(to);
+        }
+
+        function reload() {
+            location.reload();
+        }
+
+        class PalindromErrorCatcher extends HTMLElement {
+            static get observedAttributes() {
+                return ['target-selector']
+            }
+            constructor() {
+                super();
+
+                // Creates the shadow root
+                const shadowRoot = this.attachShadow({ mode: 'open' });
+                const clone = document.importNode(template.content, true);
+                shadowRoot.appendChild(clone);
+
+                this.connectionErrorPane = shadowRoot.querySelector('.connection-error-pane');
+                this.genericErrorPane = shadowRoot.querySelector('.generic-error-pane');
+                this.reconnectionSecondsSpan = shadowRoot.querySelector('.reconnection-seconds');
+                this.cancelReloadingBtn = shadowRoot.querySelector('.cancel-reloading-btn');
+                this.reloadNowBtn = shadowRoot.querySelector('.reload-btn');
+
+                bindCallbacks(this);
+            }
+            _bindToElement() {
+                this.target = this.getRootNode().querySelector(this.targetSelector) || document.querySelector(this.targetSelector);
+            }
+
+            _subscribeToEvents() {
+                this._unsubscribeToEvents();
+                this.target.addEventListener('generic-error', this.handleGenericError);
+                this.target.addEventListener('connection-error', this.handleConnectionError);
+                this.target.addEventListener('reconnection-countdown', this.handleReconnectionCountdownTick);
+                this.target.addEventListener('reconnection-end', this.handleReconnectionEnd);
+                this.cancelReloadingBtn.addEventListener('click', this.cancelReloading);
+                this.reloadNowBtn.addEventListener('click', reload);
+                this.genericErrorPane.addEventListener('click', reload);
+                this.subscribed = true;
+            }
+            _unsubscribeToEvents() {
```

Comment:
I think those three `_methods` could also be hidden for public, to make sure nobody will start using them so we will be forced to support it ;)

Reply:
But this is extremely anti-pattern. They're part of the element class and I feel leaving them outside and passing parameters to them is like making static functions in a class and using them in another class **once**. This misses the point of the class altogether. I think the underscore should suffice to say that they're not part of the API. I mean I understand doing this for utility helper functions, but this functions is like a life-cycle method for the instance, doesnt' feel right to keep it outside.

Reply:
I wouldn't say it's "extremely anti-pattern" it's opinionated, what is worse expose _private_ methods as effectively _public_ or define private methods outside of `class` keyword. There are just ways to address the problem, see more solutions at http://exploringjs.com/es6/ch_classes.html#sec_private-data-for-classes

You can see discussions like this in many places, like:
https://github.com/airbnb/javascript/issues/1024


Personally, I think putting variably/function outside of published scope [is](http://exploringjs.com/es6/ch_classes.html#sec_private-data-for-classes) and [was](http://speakingjs.com/es5/ch17.html#private_data_for_objects) **the way** to make them local/private in JavaScript since the very beginning, I don't like pretending JavaScript and its classes are like Java. Usually, I try to avoid quasi-/convention-based- private members as much as I can, unless I really need them for inheritance or smth. As I never trus convention to be reliable security guard.

Reply:
Anyway, as I said to me it's just an opinion if you have the different one, go ahead, that was just my comment and suggestion. Failing FF, and lack of contributing are more blocking problems :)